File: generated-code/haskell/codecraft/FileReadWrite/src/Model/EntityAction.hs
module Model.EntityAction where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Model.AttackAction (AttackAction)
import Model.BuildAction (BuildAction)
import Model.MoveAction (MoveAction)
import Model.RepairAction (RepairAction)

-- | Entity's action
data EntityAction = EntityAction {
    -- | Move action
    moveAction :: Maybe MoveAction,
    -- | Build action
    buildAction :: Maybe BuildAction,
    -- | Attack action
    attackAction :: Maybe AttackAction,
    -- | Repair action
    repairAction :: Maybe RepairAction }
    deriving Show

instance Trans EntityAction where
    read = do
        moveAction <- Trans.read
        buildAction <- Trans.read
        attackAction <- Trans.read
        repairAction <- Trans.read
        return EntityAction {
            moveAction,
            buildAction,
            attackAction,
            repairAction }
    
    write EntityAction {
        moveAction,
        buildAction,
        attackAction,
        repairAction } = do
            Trans.write moveAction
            Trans.write buildAction
            Trans.write attackAction
            Trans.write repairAction

File: generated-code/rust/codecraft/TcpReadWrite/src/model/entity_action.rs
use super::*;

/// Entity's action
#[derive(Clone, Debug)]
pub struct EntityAction {
    /// Move action
    pub move_action: Option<model::MoveAction>,
    /// Build action
    pub build_action: Option<model::BuildAction>,
    /// Attack action
    pub attack_action: Option<model::AttackAction>,
    /// Repair action
    pub repair_action: Option<model::RepairAction>,
}

impl trans::Trans for EntityAction {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.move_action.write_to(writer)?;
        self.build_action.write_to(writer)?;
        self.attack_action.write_to(writer)?;
        self.repair_action.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let move_action: Option<model::MoveAction> = trans::Trans::read_from(reader)?;
        let build_action: Option<model::BuildAction> = trans::Trans::read_from(reader)?;
        let attack_action: Option<model::AttackAction> = trans::Trans::read_from(reader)?;
        let repair_action: Option<model::RepairAction> = trans::Trans::read_from(reader)?;
        Ok(Self {
            move_action,
            build_action,
            attack_action,
            repair_action,
        })
    }
}